basis/basis.py

```diff
@@ -0,0 +1,172 @@
+import abc
+import numpy
+import math
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``max_exactness`` represents the maximum level of exactness, or 
+    accuracy, of a surrogate function this class can provide enough 
+    information for. This class can provide enough information for 
+    constructing a surrogate function of any accuracy at or below 
+    max_exactness.
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid. The Smolyak index of a given point is equal to 
+    its position in the list points.
+    ``levels`` property gives the Smolyak indexes for each
+    grid level.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
+        self._max_exactness = max_exactness
+        self._points = []
+        self._levels = []
+        self._update()
+
+    @property
+    def max_exactness(self):
+        """Maximum level of exactness described"""
+        return self._max_exactness
+
+    @property
+    def points(self):
+        """Extrema assigned to Smolyak indices"""
+        return self._points
+
+    @property
+    def levels(self):
+        """index of each point that belongs to each grid level"""
+        return self._levels
+
+    @max_exactness.setter
+    def max_exactness(self, max_exactness):
+        """Set maximum level of exactness
+        Sets maximmum level of exactness class instance will describe, will
+        extend or truncate properties points and levels to match new value
+
+        Parameters
+        ----------
+        max_exactness : int
+            new level of exactness class should describe
+        """
+        if max_exactness > self._max_exactness:
+            # extend
+            self._max_exactness = max_exactness
+            # update other properties
+            self._update()
+        elif max_exactness < self._max_exactness:
+            # truncate
+            self._max_exactness = max_exactness
+            # update other properties
+            points_keep = sum(
+                    [len(x) for x in self._levels[:self._max_exactness+1]])
+            self._points = self._points[:points_keep]
+            self._levels = self._levels[:self._max_exactness+1]
+
+
+    @abc.abstractmethod
+    def _update(self):
+        """Update properties described by level of exactness
+        Compute points of basis function and the Smolyak index of
+        each grid level between 0 and max_exactness + 1 
+        """
+        pass
+
+    @abc.abstractmethod
+    def basis(self,x,n):
```

Comment:
I think we should reverse the order of these arguments and convert to a call method
```suggestion
    def __call__(self, n, x):
```
so that it matches more like the math formatting `T_n(x)` like
```
T = SomeBasisFunction()
T(1, x)
```

In general, does every basis function also take `n` as an argument in order to call it? I feel like we might need to think a bit more about that.

Reply:
I thought more about this, and I think it is OK to keep n because it corresponds to one of the “point” indexes, right?

Reply:
It refers to the degree of the Chebyshev polynomial being called. In the BasisFunction class, I operated under the assumption that each basis function would be a series of related functions that are distinguished from each other by their degree like the Chebyshev polynomials or some variable analogous to a degree.

For the Chebyshev polynomials in particular, they do correspond to the point indexes. To construct the big matrix used to calculate the coefficients from slide 10 on the presentation, you have to feed the values in grid_point_indexes as n and the values of points[grid_point_indexes] as the x.

Reply:
OK, so then is “n” referring to a level or a point? It seems like a level because one Chebyshev polynomial adds multiple points, right?

Reply:
Neither really when it's used. 
The points added from one grid level are the extrema of the Chebyshev polynomial with degree `n = 2**(grid_level) + 1`
Each point then gets a Smolyak index from 0 to 2**(grid_level) + 1, and the Smolyak index of each point is reflected by its location in `points`. 
Then those Smolyak indexes get associated with a Chebyshev polynomial that has a degree equal to the Smolyak index when the surrogate function terms are being constructed. grid_point_indexes tells you the degree of the Chebyshev polynomial for each surrogate function terms.

Reply:
OK, I see, thanks! Let’s add this to the discussion list then.